rpc/provider/proxy.ts

```diff
@@ -1,95 +1,107 @@
-import * as U from "../../util/mod.ts"
+import { retry } from "../../deps/std/async.ts"
 import * as msg from "../messages.ts"
-import { nextIdFactory, Provider, ProviderConnection, ProviderListener } from "./base.ts"
+import { ListenersContainer, nextIdFactory, Provider, ProviderListener } from "./base.ts"
 import { ProviderCloseError, ProviderHandlerError, ProviderSendError } from "./errors.ts"
 
-/** Global lookup of existing connections */
-const connections = new Map<string, ProxyProviderConnection>()
-type ProxyProviderConnection = ProviderConnection<WebSocket, Event, Event>
-
 const nextId = nextIdFactory()
 
+const listenersContainer = new ListenersContainer<string, Event, Event>()
+const activeWs = new Map<string, WebSocket>()
+const connectingWs = new Map<string, Promise<WebSocket>>()
+const CUSTOM_WS_CLOSE_CODE = 4000
+
 export const proxyProvider: Provider<string, Event, Event, Event> = (url, listener) => {
```

Comment:
I believe that we need to pass [retryOptions](https://deno.land/std@0.170.0/async/mod.ts?s=RetryOptions) here.
Perhaps, creating with a `proxyProviderFactory`
```
const proxyProviderFactory = (retryOptions) => (url, listener) => {...};
const proxyProvider = proxyProviderFactory(defaultRetryOptions);
```

Thoughts?